### DISCUSSION TITLE
File upload

### DISCUSSION THREAD

**Discussion by bq-wrongway** [pocketbase/pocketbase#493] [Q&A]
Hello, 
i am using flutter and i was playing around with the pocketbase for couple of days now, i stumbled upon an issue that i dont know how to solve, since i get no error or any feedback on whats happening. 
Using this https://pocketbase.io/docs/files-handling/ , i setup literaly the same scenario my collection is called imagefiles, and it has two fields, field file of **type file** and field title of **type string**.

When i try to perform the upload, the title gets uploaded but the file does not (it does not mater if its text file, image file, if the path is absolute or relative), i never get any error, or any warning it just does nothing to the file. 

`void uploadImage() async {
    await client.admins.authViaEmail('user@thu.com', 'c24jjj55pA');
    final record = await client.records.create(
 'imagefiles',
      body: {
        'title': 'new img!', // some regular text field
      },
      files: [
        http.MultipartFile.fromString(
          'file[0]',
          'demo content 2...',
          filename: '/home/username/Downloads/test.jpg',
        ),
      ],
    );
  }`

I am using http for dart 0.13.5

Could anyone give me a hand, if i had any error i would know what to look for , like this i am just lost. 
Could it be some permission issue on linux ? (but that folder does not have any restrictions)

**Comment by ganigeorgiev** [comment]
@bqMoonlight I think you are using an incorrect file fied name when generating the `http.MultipartFile` instance. Instead of `file[0]` pass the name of file field that you want to update as shown in the [Dart SDK example](https://github.com/pocketbase/dart-sdk#file-upload) (_the site documentation is a little confusing, I'll try to update it later_).

For debugging you can add the `--debug` flag when starting PocketBase and/or check the logs in the Admin UI.

_I'll move the issue to Q&A discussion._

**Comment by bq-wrongway** [comment]
Well i managed to do it , yes i made a mistake when naming a file, but uploading image file must be done with from bytes ? we have to convert image to bytes and then do it, since i think that using that string option just appends that string to whatever the underlaying file is ?

**Reply by ganigeorgiev** [reply]
Yes, `http.MultipartFile.fromString(...)` is just an example. You can use any other factory method that suite your use case as long as it results to a `http.MultipartFile` instance.

**Reply by ganigeorgiev** [reply]
@bqMoonlight There is also a `fromPath()` static method that you could use instead of manually converting the image to bytes - https://pub.dev/documentation/http/latest/http/MultipartFile/fromPath.html (note that it resolves to future, so you'll need to await it).

**Comment by bq-wrongway** [comment]
But from path would not work on web right ? 
I was just using flutter for web like a playground.

**Reply by ganigeorgiev** [reply]
Yes, the documentaiton state that `fromPath` is supported only for `dart:io` environments:
> Throws an [UnsupportedError](https://api.dart.dev/stable/2.18.0/dart-core/UnsupportedError-class.html) if dart:io isn't supported in this environment.

The only requirement in the PocketBase Dart SDK is that the file must be a `http.MultipartFile` instance.
How this instance was created depends on your use case and you are free to use any available method for that.

**Comment by bq-wrongway** [comment]
Yes i managed to get this what i was going, but later i dont understand how you actually delete file, when there are multiple files in record.

 i retrieve the files like this 
 
```
Future<List<Uri>> getImageFromStorage() async {
    final recordX =
        await client.records.getOne('imagefiles', 'recordId');
    recordX.data['files'].forEach((e) {
      list.add(client.records.getFileUrl(recordX, e));
    });

    return list;
  }
```

But if i want to delete one file in the file list from a record, i should update the record and set that particular item to null,  and while in javascript example it looks simple, because you just set certain index to null, dart counterpart seems more complex, since it accepts only Multipart list. (does that mean i should send whole list back with null value on desired index ?)
This is something i still did not figure out how to do.
I would appreciate some help, il give it a shot tomorrow again, i am ill and exausted, so i might even not see something obvious.

**Reply by ganigeorgiev** [reply]
Deleting single file from a multiple upload field It is the same as in the JS SDK. You have to set `field.index: null` as body param, eg.:

```dart
await client.records.update('demo', 'RECORD_ID',
    body: {
        'files': null, // delete all files
        'files.0': null, // delete only the first file
    },
})
```

The Multipart list is only for upload. I'll consider adding Dart example when refactoring the documentation.

**Reply by ganigeorgiev** [reply]
@bqMoonlight Just to let you know that I've updated the existing "Files upload and handling" docs to have Dart equivalent examples - https://pocketbase.io/docs/files-handling/.

**Reply by bq-wrongway** [reply]
Thanks a lot man, this worked. 
 I mean you answered my original question and then some, very kind. 
 From my side this matter is closed!
